### PR TITLE
use segregated hmap to boost the freelist allocate and release performace

### PR DESCRIPTION
**Pull request by WIZARD-CXY**
In this pr, I use segregated bitmap to replace the original freeids allocating and releasing approach.
It is much faster than the original version, especially when the db size is large or the fragmentation in the db is large, we can gain 1000x faster performance.

### CODE REVIEW

**Comment by WIZARD-CXY** [comment]
@xiang90 ptal

File: freelist.go [hunk 1]
```diff
@@ -160,8 +277,98 @@ func (f *freelist) release(txid txid) {
...
-	f.ids = pgids(f.ids).merge(m)
+
+	for _, pgid := range m {
+		// try to see if we can merge and update
+		f.mergeWithExistingSpan(pgid)
+	}
+}
+
+// mergeWithExistingSpan merges pgi to the existing free spans, try to merge it backward and forward
+func (f *freelist) mergeWithExistingSpan(pgi pgid) {
```
Comment (hormes) [inline freelist.go]: this function can be written like:
start, size, pgi := ...
if prev {
    remove prev span
    start, size, pgi := ...
}
if next {
    remove next span
    start, size, pgi := ...
}
add this span ...

Comment (WIZARD-CXY) [reply freelist.go]: @hormes pal

Comment (WIZARD-CXY) [reply freelist.go]: pedal

Comment (WIZARD-CXY) [reply freelist.go]: pedal

**Comment by WIZARD-CXY** [comment]
@benbjohnson

**Comment by mitake** [comment]
@WIZARD-CXY Interesting! Could you share how you measure the performance improvement?

File: freelist.go [hunk 2]
```diff
@@ -13,22 +13,69 @@ type txPending struct {
...
+	f.forwardMap = make(map[pgid]uint64)
+	f.backwardMap = make(map[pgid]uint64)
+	if len(pgids) == 0 {
+		return
+	}
+	if !sort.SliceIsSorted([]pgid(pgids), func(i, j int) bool { return pgids[i] < pgids[j] }) {
+		panic("pgids not sorted")
+	}
+
+	if len(pgids) == 1 {
```
Comment (xiang90) [inline freelist.go]: do we really need this special case handling?

File: freelist.go [hunk 3]
```diff
@@ -49,7 +96,12 @@ func (f *freelist) count() int {
 
 // free_count returns count of free pages
 func (f *freelist) free_count() int {
-	return len(f.ids)
+	//use the forwardmap to get the total count
```
Comment (xiang90) [inline freelist.go]: how often will this func be called? probably consider add a counter to record the number of free pages.

File: freelist.go [hunk 4]
```diff
@@ -49,7 +96,12 @@ func (f *freelist) count() int {
...
+	res := 0
```
Comment (xiang90) [inline freelist.go]: count := 0. do not use very generic variable naming like res.

File: freelist.go [hunk 5]
```diff
@@ -69,50 +121,82 @@ func (f *freelist) copyall(dst []pgid) {
 		m = append(m, txp.ids...)
 	}
 	sort.Sort(m)
-	mergepgids(dst, f.ids, m)
+	mergepgids(dst, f.getFreeids(), m)
+}
+
+// getFreeids returns the sorted free page ids
+func (f *freelist) getFreeids() []pgid {
```
Comment (xiang90) [inline freelist.go]: getFreePageIDs

File: freelist.go [hunk 6]
```diff
@@ -69,50 +121,82 @@ func (f *freelist) copyall(dst []pgid) {
...
+	if f.free_count() == 0 {
```
Comment (xiang90) [inline freelist.go]: no need to add this short cut.

File: freelist.go [hunk 7]
```diff
@@ -13,22 +13,69 @@ type txPending struct {
 	alloctx          []txid // txids allocating the ids
 	lastReleaseBegin txid   // beginning txid of last matching releaseRange
 }
+type pidSet map[pgid]struct{}
```
Comment (xiang90) [inline freelist.go]: add comment.

File: freelist.go (hunk 4 above)
Comment (WIZARD-CXY) [reply freelist.go]: ok

File: freelist.go (hunk 2 above)
Comment (WIZARD-CXY) [reply freelist.go]: no need

File: freelist.go (hunk 7 above)
Comment (WIZARD-CXY) [reply freelist.go]: ok

File: freelist.go [hunk 8]
```diff
@@ -69,50 +121,82 @@ func (f *freelist) copyall(dst []pgid) {
...
+	sort.Sort(pgids(m))
+
+	return m
 }
 
 // allocate returns the starting page id of a contiguous list of pages of a given size.
 // If a contiguous block cannot be found then 0 is returned.
 func (f *freelist) allocate(txid txid, n int) pgid {
-	if len(f.ids) == 0 {
+	if n == 0 {
```
[the remark this answers is not in the document]
Comment (WIZARD-CXY) [reply freelist.go]: yeah, common use case we don't request a 0-size page, this is only wanted by unit test in allocating
```
f.allocate(1, 0)
	if x := f.free_count(); x != 2 {
		t.Fatalf("exp=2; got=%v", x)
	}
```

File: freelist.go [hunk 9]
```diff
@@ -160,8 +244,60 @@ func (f *freelist) release(txid txid) {
...
+func (f *freelist) mergeWithExistingSpan(pid pgid) {
+	prev := pid - 1
+	next := pid + 1
+
+	size1, ok1 := f.backwardMap[prev]
+	size2, ok2 := f.forwardMap[next]
+	newStart := pid
+	newSize := uint64(1)
+
+	if ok1 {
```
[the remark this answers is not in the document]
Comment (WIZARD-CXY) [reply freelist.go]: ok

File: freelist.go [hunk 10]
```diff
@@ -69,50 +121,82 @@ func (f *freelist) copyall(dst []pgid) {
...
+	// if we have a exact size match just return short path
+	if bm, ok := f.freemaps[uint64(n)]; ok {
+		if len(bm) == 0 {
+			delete(f.freemaps, uint64(n))
+		} else {
+			for pid := range bm {
+				// remove the span
+				f.delSpan(pid, uint64(n))
+
+				f.allocs[pid] = txid
```
[the remark this answers is not in the document]
Comment (WIZARD-CXY) [reply freelist.go]: no need, just the start of the span is enough. The original approach is the same.

**Comment by WIZARD-CXY** [comment]
@mitake  we use bboltdb in etcd. In the original approach, when we have a very large db(~50GB), and do the put op(~5000op/s), we find the db spill time is very large(~8s), the put latency is incredibly large too. We look into the code, add some debugging info and find that freelist allocation is the bottleneck.It is because when the freelist size is large and its inner fragmentation is very common. The original algorithm in `allocate` tries so hard to get a start page id while in my approach we use a hash map to hold the same size of starting pgids. So the allocation is super fast(nearly O(1)) and the release performance is boosted too(also o(1), removed the original page id sort operation)

**Comment by WIZARD-CXY** [comment]
@mitake https://github.com/boltdb/bolt/issues/640 this is the same case

**Comment by xiang90** [comment]
@WIZARD-CXY Let us make this an option instead of entirely replacing the array based freelist implementation.

Probably add an option called FreelistType: (array/seglist)

File: freelist.go [hunk 11]
```diff
@@ -49,7 +91,12 @@ func (f *freelist) count() int {
 
 // free_count returns count of free pages
 func (f *freelist) free_count() int {
-	return len(f.ids)
+	//use the forwardmap to get the total count
```
Comment (xiang90) [inline freelist.go]: need an empty space after `//`

File: freelist.go [hunk 12]
```diff
@@ -14,21 +14,63 @@ type txPending struct {
...
+	}
+}
+
+// initial from 0
+// pgids must be sorted
+func (f *freelist) init(pgids []pgid) {
+	f.freemaps = make(map[uint64]pidSet)
+	f.forwardMap = make(map[pgid]uint64)
+	f.backwardMap = make(map[pgid]uint64)
+	if len(pgids) == 0 {
```
Comment (xiang90) [inline freelist.go]: remove this if checking. it is not necessary.

File: freelist.go [hunk 13]
```diff
@@ -245,15 +367,9 @@ func (f *freelist) read(p *page) {
...
-	if count == 0 {
-		f.ids = nil
-	} else {
+	if count != 0 {
 		ids := ((*[maxAllocSize]pgid)(unsafe.Pointer(&p.ptr)))[idx : idx+count]
-		f.ids = make([]pgid, len(ids))
-		copy(f.ids, ids)
-
-		// Make sure they're sorted.
-		sort.Sort(pgids(f.ids))
```
Comment (xiang90) [inline freelist.go]: why the sorting is removed?

File: freelist.go [hunk 14]
```diff
@@ -262,7 +378,7 @@ func (f *freelist) read(p *page) {
 
 // read initializes the freelist from a given list of ids.
 func (f *freelist) readIDs(ids []pgid) {
-	f.ids = ids
+	f.init(ids)
```
Comment (xiang90) [inline freelist.go]: hm... how do we know the passed in ids is sorted?

File: freelist.go (hunk 12 above)
Comment (WIZARD-CXY) [reply freelist.go]: this necessary, see line 57

Comment (xiang90) [reply freelist.go]: oh. ok. you are right.

**Comment by WIZARD-CXY** [comment]
>
@WIZARD-CXY Let us make this an option instead of entirely replacing the array based freelist implementation.

Probably add an option called FreelistType: (array/seglist)

working on it

**Comment by WIZARD-CXY** [comment]
@mitake @xiang90 @hormes petal add a commit to add an option of using the new approach.

File: db.go [hunk 15]
```diff
@@ -70,6 +70,13 @@ type DB struct {
 	// re-sync during recovery.
 	NoFreelistSync bool
 
+	// IsFreelistTypeArray sets the backend freelist type to array which is simple but endures
+	// dramatic performance degradation if database is large and framentation in freelist is common.
+	// The alternative one is using segregated freelist, it is faster in almost all circumstances
+	// but it doesn't guarantee that it offers the smallest page id available. In normal case it is safe.
+	// if you depend on this(it offers the smallest page id available), you may set this to true.
+	IsFreelistTypeArray bool
```
Comment (xiang90) [inline db.go]: can we make this an option rather than a boolean typed selector?

File: freelist.go [hunk 16]
```diff
@@ -69,12 +94,18 @@ func (f *freelist) copyall(dst []pgid) {
...
-	mergepgids(dst, f.ids, m)
+
+	mergepgids(dst, f.getFreePageIDs(), m)
 }
 
 // allocate returns the starting page id of a contiguous list of pages of a given size.
 // If a contiguous block cannot be found then 0 is returned.
 func (f *freelist) allocate(txid txid, n int) pgid {
+	if !f.IsFreelistTypeArray {
+
```
Comment (xiang90) [inline freelist.go]: remove empty line.

File: freelist.go [hunk 17]
```diff
@@ -49,6 +70,10 @@ func (f *freelist) count() int {
 
 // free_count returns count of free pages
 func (f *freelist) free_count() int {
+	// use the forwardmap to get the total count
```
Comment (xiang90) [inline freelist.go]: move this comment inside the if statement.

File: freelist.go [hunk 18]
```diff
@@ -151,6 +182,12 @@ func (f *freelist) free(txid txid, p *page) {
 
 // release moves all page ids for a transaction id (or older) to the freelist.
 func (f *freelist) release(txid txid) {
+	if !f.IsFreelistTypeArray {
+		f.seglistRelease(txid)
+
```
Comment (xiang90) [inline freelist.go]: remove this empty line.

File: freelist.go [hunk 19]
```diff
@@ -166,10 +203,16 @@ func (f *freelist) release(txid txid) {
 
 // releaseRange moves pending pages allocated within an extent [begin,end] to the free list.
 func (f *freelist) releaseRange(begin, end txid) {
+	if !f.IsFreelistTypeArray {
+		f.seglistReleaseRange(begin, end)
+
```
Comment (xiang90) [inline freelist.go]: remove this empty line.

File: freelist.go [hunk 20]
```diff
@@ -200,6 +243,11 @@ func (f *freelist) releaseRange(begin, end txid) {
 
 // rollback removes the pages from a given pending tx.
 func (f *freelist) rollback(txid txid) {
+	if !f.IsFreelistTypeArray {
+		f.seglistRollback(txid)
+
```
Comment (xiang90) [inline freelist.go]: remove this empty line.

File: freelist.go [hunk 21]
```diff
@@ -233,6 +281,11 @@ func (f *freelist) freed(pgid pgid) bool {
 
 // read initializes the freelist from a freelist page.
 func (f *freelist) read(p *page) {
+	if !f.IsFreelistTypeArray {
+		f.seglistRead(p)
+
```
Comment (xiang90) [inline freelist.go]: remove this empty line.

File: db.go (hunk 15 above)
Comment (WIZARD-CXY) [reply db.go]: the old options are boolean-like(like NoFreelistSync)... I think it is ok to keep this way

File: allocate_test.go [hunk 22]
```diff
@@ -6,7 +6,12 @@ import (
 
 func TestTx_allocatePageStats(t *testing.T) {
 	f := newFreelist()
-	f.ids = []pgid{2, 3}
+	if !f.IsFreelistTypeArray {
+		ids := []pgid{2, 3}
```
Comment (xiang90) [inline allocate_test.go]: move this out of the if/else statement.

File: allocate_test.go [hunk 23]
```diff
@@ -6,7 +6,12 @@ import (
...
+		f.init(ids)
+	} else {
+		f.ids = []pgid{2, 3}
```
Comment (xiang90) [inline allocate_test.go]: actually you can also add a f.init func to the array type freelist. so we do not need this if/else anymore.

File: freelist.go (hunk 13 above)
Comment (WIZARD-CXY) [reply freelist.go]: no need

File: allocate_test.go (hunk 23 above)
Comment (WIZARD-CXY) [reply allocate_test.go]: changed by using `readIDs`

File: allocate_test.go (hunk 22 above)
Comment (WIZARD-CXY) [reply allocate_test.go]: fixed

File: .gitignore [hunk 24]
```diff
@@ -2,3 +2,4 @@
 *.test
 *.swp
 /bin/
+cover.out
```
Comment (xiang90) [inline .gitignore]: separate this into another pull request.

**Comment by xiang90** [comment]
@WIZARD-CXY Can you create a new PR? It can be cleaner.

**Comment by mitake** [comment]
@WIZARD-CXY thanks for sharing the detail and sorry for my delayed reply. It's really interesting. Do you mean you already have etcd cluster with 50GB of data? Then are you seeing availability problem comes from the large snapshot?
I'm asking for understanding your use case and for curious: I heard alibaba has very large clusters. Is this related that?

**Comment by WIZARD-CXY** [comment]
@mitake yeah, we stored many data in etcd since we have large clusters
